File: examples/java/src/main/java/io/superdurable/dex/patterns/parallelsubflows/WaitForHalfParentFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

package io.superdurable.dex.patterns.parallelsubflows;

import io.superdurable.dex.Channel;
import io.superdurable.dex.Client;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.FlowResult;
import io.superdurable.dex.FlowStatus;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.SubFlow;
import io.superdurable.dex.Wait;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.stereotype.Component;

@Component
public final class WaitForHalfParentFlow implements Flow<String[]> {
    public final Channel<Boolean> subFlowCompletedCh =
            Channel.define("SubFlowCompletedCh", Boolean.class);
    public final Channel<Boolean> allDoneCh = Channel.define("AllDoneCh", Boolean.class);

    private final InitStep initStep = new InitStep();
    private final SubFlowStep subFlowStep;
    private final WaitSubFlowsStep waitSubFlowsStep = new WaitSubFlowsStep();

    public WaitForHalfParentFlow(final ObjectProvider<Client> clientProvider) {
        subFlowStep = new SubFlowStep(clientProvider);
    }

    @Override
    public StepList<String[]> getSteps() {
        return StepList.startStep(initStep).otherSteps(subFlowStep, waitSubFlowsStep);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(subFlowCompletedCh, allDoneCh);
    }

    final class InitStep implements Step<String[]> {
        @Override
        public Class<String[]> getInputType() {
            return String[].class;
        }

        @Override
        public StepDecision execute(final Context context, final String[] requests) {
            if (requests.length == 0) {
                return StepDecision.gracefulComplete(null);
            }
            final StepMovement<?>[] movements = new StepMovement<?>[requests.length + 1];
            movements[0] = StepMovement.of(WaitSubFlowsStep.class, requests.length);
            for (int index = 0; index < requests.length; index++) {
                movements[index + 1] = StepMovement.of(SubFlowStep.class, requests[index]);
            }
            return StepDecision.goToMany(movements);
        }
    }

    final class SubFlowStep implements Step<String> {
        private final ObjectProvider<Client> clientProvider;
        SubFlowStep(final ObjectProvider<Client> clientProvider) {
            this.clientProvider = clientProvider;
        }

        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public Wait waitFor(final Context context, final String request) {
            return Wait.anyOf(SubFlow.run(ExampleSubFlow.class, request), allDoneCh.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final String request) {
            final FlowResult result = SubFlow.getConditionResults(context);
            if (result.getStatus() != FlowStatus.RUNNING) {
                subFlowCompletedCh.publish(context, true);
                return StepDecision.gracefulComplete(null);
            }
            clientProvider.getObject().stopFlow(SubFlow.getFlowId(context));
            return StepDecision.gracefulComplete(null);
        }
    }

    final class WaitSubFlowsStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer total) {
            return Wait.until(subFlowCompletedCh.forN((total + 1) / 2));
        }

        @Override
        public StepDecision execute(final Context context, final Integer total) {
            final int remaining = total - (total + 1) / 2;
            for (int index = 0; index < remaining; index++) {
                allDoneCh.publish(context, true);
            }
            return StepDecision.gracefulComplete(null);
        }
    }
}

File: examples/go/patterns/parallel-subflows/wait_for_half_parent_flow.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package parallelsubflows

import (
	"context"
	"fmt"

	"github.com/superdurable/dex/sdk-go/dex"
)

var (
	SubFlowCompletedCh = dex.DefineChannel[bool]("SubFlowCompletedCh")
	AllDoneCh          = dex.DefineChannel[bool]("AllDoneCh")
)

type WaitForHalfParentFlow struct {
	dex.FlowDefaults
	getClient   func() *dex.Client
	exampleFlow *ExampleSubFlow
}

func NewWaitForHalfParentFlow(
	getClient func() *dex.Client,
	exampleFlow *ExampleSubFlow,
) *WaitForHalfParentFlow {
	return &WaitForHalfParentFlow{getClient: getClient, exampleFlow: exampleFlow}
}

func (flow *WaitForHalfParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{
		dex.DefineStartStep(waitForHalfInit{}),
		dex.DefineStep(subFlow{getClient: flow.getClient, exampleFlow: flow.exampleFlow}),
		dex.DefineStep(waitSubFlows{}),
	}
}

func (*WaitForHalfParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{Channels: []dex.ChannelDef{SubFlowCompletedCh, AllDoneCh}}
}

type waitForHalfInit struct {
	dex.StepDefaultsNoWaitFor[[]string]
}

func (waitForHalfInit) GetStepType() string { return "Init" }

func (waitForHalfInit) Execute(_ dex.Context, requests []string) (*dex.StepDecision, error) {
	if len(requests) == 0 {
		return dex.GracefulComplete(nil), nil
	}
	movements := make([]dex.StepMovement, 0, len(requests)+1)
	movements = append(movements, dex.MovementOf(waitSubFlows{}, len(requests)))
	for _, request := range requests {
		movements = append(movements, dex.MovementOf(subFlow{}, request))
	}
	return dex.GoToMany(movements...), nil
}

type subFlow struct {
	dex.StepDefaults
	getClient   func() *dex.Client
	exampleFlow *ExampleSubFlow
}

func (subFlow) GetStepType() string { return "SubFlow" }

func (step subFlow) WaitFor(_ dex.Context, request string) (*dex.Wait, error) {
	return dex.AnyOf(dex.SubFlow(step.exampleFlow, request), AllDoneCh.ForOne()), nil
}

func (step subFlow) Execute(ctx dex.Context, _ string) (*dex.StepDecision, error) {
	result, err := dex.SubFlowResult(ctx)
	if err != nil {
		return nil, err
	}
	if result.Status != dex.FlowRunning {
		if err := SubFlowCompletedCh.Publish(ctx, true); err != nil {
			return nil, err
		}
		return dex.GracefulComplete(nil), nil
	}
	client := step.getClient()
	if client == nil {
		return nil, fmt.Errorf("dex client is not available")
	}
	flowID, err := dex.SubFlowID(ctx)
	if err != nil {
		return nil, err
	}
	if err := client.StopFlow(context.Background(), flowID, dex.StopOptions{
		Type: dex.CancelFlow, Reason: "enough SubFlows completed",
	}); err != nil {
		return nil, err
	}
	return dex.GracefulComplete(nil), nil
}

type waitSubFlows struct{ dex.StepDefaults }

func (waitSubFlows) GetStepType() string { return "WaitSubFlows" }

func (waitSubFlows) WaitFor(_ dex.Context, total int) (*dex.Wait, error) {
	return dex.Until(SubFlowCompletedCh.ForN((total + 1) / 2)), nil
}

func (waitSubFlows) Execute(ctx dex.Context, total int) (*dex.StepDecision, error) {
	remaining := total - (total+1)/2
	for index := 0; index < remaining; index++ {
		if err := AllDoneCh.Publish(ctx, true); err != nil {
			return nil, err
		}
	}
	return dex.GracefulComplete(nil), nil
}

var _ dex.Flow = (*WaitForHalfParentFlow)(nil)
